src/containers/scrollable-canvas.jsx

```diff
@@ -0,0 +1,174 @@
+import paper from '@scratch/paper';
+import PropTypes from 'prop-types';
+
+import React from 'react';
+import {connect} from 'react-redux';
+import ScrollableCanvasComponent from '../components/scrollable-canvas/scrollable-canvas.jsx';
+
+import {ART_BOARD_WIDTH, ART_BOARD_HEIGHT, clampViewBounds, pan, zoomOnFixedPoint} from '../helper/view';
+import {updateViewBounds} from '../reducers/view-bounds';
+import {redrawSelectionBox} from '../reducers/selected-items';
+
+import {getEventXY} from '../lib/touch-utils';
+import bindAll from 'lodash.bindall';
+
+class ScrollableCanvas extends React.Component {
+    static get ZOOM_INCREMENT () {
+        return 0.5;
+    }
+    constructor (props) {
+        super(props);
+        bindAll(this, [
+            'handleHorizontalScrollbarMouseDown',
+            'handleHorizontalScrollbarMouseMove',
+            'handleHorizontalScrollbarMouseUp',
+            'handleVerticalScrollbarMouseDown',
+            'handleVerticalScrollbarMouseMove',
+            'handleVerticalScrollbarMouseUp',
+            'handleWheel'
+        ]);
+    }
+    componentDidMount () {
+        if (this.props.canvas) {
+            this.props.canvas.addEventListener('wheel', this.handleWheel);
+        }
+    }
+    componentWillReceiveProps (nextProps) {
+        if (nextProps.canvas) {
+            if (this.props.canvas) {
+                this.props.canvas.removeEventListener('wheel', this.handleWheel);
+            }
+            nextProps.canvas.addEventListener('wheel', this.handleWheel);
+        }
+    }
+    handleHorizontalScrollbarMouseDown (event) {
+        window.addEventListener('mousemove', this.handleHorizontalScrollbarMouseMove);
+        window.addEventListener('mouseup', this.handleHorizontalScrollbarMouseUp);
+        this.initialMouseX = getEventXY(event).x;
+        this.initialScreenX = paper.view.matrix.tx;
```

Comment:
Should this be set before binding the event listeners? I'm wondering if it's possible to have a race condition where `handleHorizontalScrollbarMouseMove` is called before `this.initialMouseX` or `this.initialScreenX` are set...

Reply:
I think it's fine since JS is single-threaded, but I switched them for logical clarity